docs(mdx): add a dynamic imports section for App Router (#73466)
vercel/next.js: docs/01-app/02-building-your-application/07-configuring/05-mdx.mdx

Navigating to the `/mdx-page` route should display your rendered MDX page.

<AppOnly>

### Using dynamic imports

> You can also dynamically import MDX in your pages.

Reviewer: It would be good to explain why you might want to import MDX dynamically.

Reviewer 2: Yeah- in my case the markdown didn't live in the app directory, but I still wanted the markdown to be part of the Next.js import graph so I could reap all benefits - specifically images.

Author: > in my case the markdown didn't live in the app directory, but I still wanted the markdown to be part of the Next.js import graph so I could reap all benefits - specifically images.

I guess this is a benefit, but not necessarily because of the ability to import MDX dynamically?

Reviewer 3: I think having the content outside of the `app/` directory will be one of the most common reasons for users to look for this functionality

Maybe other benefits could be:

1. Programmatically importing some content (eg. programmatic logic beyond a simple dynamic route segment like `[slug]/page.mdx)
2. Importing MDX fragments smaller than a full page
3. Enabling greater control over the layers in between the page and the MDX content (eg. via wrapping intermediary React components)

Author: I'll mix in point 1, but 2-3 is also possible without the use of a dynamic import.

Create a new dynamic page within the `/app` directory and add MDX files wherever you'd like:

```txt
  my-project
  ├── app
  │   └── blog
  │       └── [slug]
  │           └── page.(tsx/js)
  ├── markdown
  │   ├── welcome.(mdx/md)
      └── about.(mdx/md)
  |── mdx-components.(tsx/js)
  └── package.json

> ```

Reviewer: @samcx should we make a Figma diagram for this structure? Happy to help 😄

Author: That would be great!

Dynamically import the MDX file inside the page and add a [`generateStaticParams`](/docs/app/api-reference/functions/generate-static-params) to statically prerender the pages with the MDX content:

```tsx filename="app/blog/[slug]/page.tsx" switcher
export default async function Page({
  params,
}: {
  params: Promise<{ slug: string }>
}) {
  const slug = (await params).slug

>   const { default: Post } = await import(`@/markdown/${slug}.mdx`)

Reviewer 2: needs a comment on how webpack dynamic importing works - I specifically was importing too broad:

`import('@${mdxPath}'` which was causing webpack to index every single file in the repo, and webpack was running out of memory. 

The static string of `@/markdown` is very important here, as is the static string of `.mdx` extension as otherwise it will find other files in that directory and can run into issues if it doesn't know how to import them. 

I have .mp4 files alongside my blog posts and it was hitting an issue where it didn't know how to handle mp4 files.

Reviewer 4: Helpful, thank you!

  │           └── page.(tsx/js)
  ├── content
  │   ├── welcome.(mdx/md)
      └── about.(mdx/md)
  |── mdx-components.(tsx/js)
  └── package.json
```

Dynamically import the MDX file inside the page and add a [`generateStaticParams`](/docs/app/api-reference/functions/generate-static-params) to statically prerender the pages with the MDX content:

```tsx filename="app/blog/[slug]/page.tsx" switcher
export default async function Page({
  params,
}: {
  params: Promise<{ slug: string }>
}) {
  const slug = (await params).slug
  const { default: Post } = await import(`@/content/${slug}.mdx`)

  return <Post />
}

export function generateStaticParams() {
  return [{ slug: 'welcome' }, { slug: 'about' }]
}

export const dynamicParams = false
```

```jsx filename="app/blog/[slug]/page.js" switcher
export default async function Page({ params }) {
  const slug = params.slug
  const { default: Post } = await import(`@/content/${slug}.mdx`)

  return <Post />
}

export function generateStaticParams() {
  return [{ slug: 'welcome' }, { slug: 'about' }]
}

export const dynamicParams = false
```

Navigating to `/blog/welcome` or `/blog/about` should display your prerendered MDX page.

> > **Good to know**: Make sure to include the relative path (e.g., `@/content`) and `.mdx` file extension in your `import()` so Webpack does not attempt to index every file.

Reviewer 3: webpack is lowercase

```suggestion
> **Good to know**: Make sure to include the relative path (e.g., `@/content`) and `.mdx` file extension in your `import()` so webpack does not attempt to index every file.
```

Navigating to the `/mdx-page` route should display your rendered MDX page.

<AppOnly>

### Using dynamic imports

> You can also dynamically import MDX in your pages. This is useful for:
> 
> - Programmatically rendering MDX.
> - Pairing it with dynamic route segments and [`generateStaticParams`](/docs/app/api-reference/functions/generate-static-params).
> 
> Create a new dynamic page within the `/app` directory and add MDX files wherever you'd like:

Reviewer 4: ```suggestion
You can import dynamic MDX components instead of using filesystem routing for MDX files.

For example, you can have a dynamic route segment which loads MDX components from a separate directory:
```

Reviewer 3: @leerob Oh, was the suggestion here not good for some reason? It originally came from this discussion with @delbaoliveira , including my comment about "Programmatically importing some content (eg. programmatic logic beyond a simple Dynamic Segment like `[slug]/page.mdx`)":

- https://github.com/vercel/next.js/pull/73466#discussion_r1867546629

Maybe the bullet above is too vague / obscure / jargony in your opinion? I can understand that "programmatically" can be jargony and abstract.

To be clear, the pattern that I was suggesting is something like this:

`app/appointments/[slug]/page.tsx`

```tsx
type Props = {
  params: Promise<{
    slug: string;
  }>;
};

export default function AppointmentPage(props: Props) {
  let MDXContent;

  // Simplified example
  if (isLecture()) {
    MDXContent = (await import(`@/content/lectures/${slug}/index.mdx`)).default;
  } else {
    MDXContent = (await import(`@/content/appointments/${slug}/index.mdx`)).default;
  }

  return <MDXContent />;
}
```

I guess another way of expressing this particular use case would be "conditional MDX imports" - although I guess that there may be other "programmatic" use cases.

Reviewer 4: Programatically felt unnecessary to explain to the code. "dynamically" here should accomplish the same goal.

```txt
  my-project
  ├── app
  │   └── blog
  │       └── [slug]
  │           └── page.(tsx/js)
  ├── content
  │   ├── welcome.(mdx/md)
      └── about.(mdx/md)
  |── mdx-components.(tsx/js)
  └── package.json
```

> Dynamically import the MDX file inside the page and add a `generateStaticParams` to statically prerender the pages with the MDX content:

Reviewer 4: ```suggestion
[`generateStaticParams`](/docs/app/api-reference/functions/generate-static-params) can be used to prerender the provided routes. By marking `dynamicParams` as `false`, accessing a route not defined in `generateStaticParams` will 404.
```

```tsx filename="app/blog/[slug]/page.tsx" switcher
export default async function Page({
  params,
}: {
  params: Promise<{ slug: string }>
}) {
  const slug = (await params).slug
  const { default: Post } = await import(`@/content/${slug}.mdx`)

  return <Post />
}

export function generateStaticParams() {
  return [{ slug: 'welcome' }, { slug: 'about' }]
}

export const dynamicParams = false
```

```jsx filename="app/blog/[slug]/page.js" switcher
export default async function Page({ params }) {
  const slug = params.slug
  const { default: Post } = await import(`@/content/${slug}.mdx`)

  return <Post />
}

export function generateStaticParams() {
  return [{ slug: 'welcome' }, { slug: 'about' }]
}

export const dynamicParams = false
```

> Navigating to `/blog/welcome` or `/blog/about` should display your prerendered MDX page.
> 
> > **Good to know**: Make sure to include the relative path (e.g., `@/content`) and `.mdx` file extension in your `import()` so webpack does not attempt to index every file.

Reviewer 4: ```suggestion
> **Good to know**: Ensure you specify the `.mdx` file extension in your import. While it is not required to use [module path aliases](/docs/app/getting-started/installation#set-up-absolute-imports-and-module-path-aliases) (e.g., `@/content`), it does simplify your import path.
```
